### ISSUE TITLE
Backstabbing Penalty - Design Choice Question

### ISSUE THREAD

**Issue by HeathcliffWarriors** [LoneGazebo/Community-Patch-DLL#5520]
_1. Mod version (i.e Date - 4/23):_ 7/17-3

I'm reviewing the code in CvTeam.cpp in order to understand DoPlayerDeclaredWarOnSomeone in the diplomacy AI, and something seems odd to me, so I wanted to ask about why a choice was made.
```
		//are we an aggressor?
		//These only trigger on the 'first' major declaration - no chain declarations.
		if (bAggressor && !bDefensivePact)
		{
			//DPs and Vassal civs - we have to DOW them as well!
			for (iI = 0; iI < MAX_TEAMS; iI++)
			{
				if (!GET_TEAM((TeamTypes)iI).isAlive())
					continue;
				if (GET_TEAM((TeamTypes)iI).isMinorCiv())
					continue;

				//Defensive pacts and vassals trigger here.
				if (GET_TEAM((TeamTypes)iI).IsHasDefensivePact(eTeam) || GET_TEAM((TeamTypes)iI).IsVassal(eTeam))
				{
#if defined(MOD_EVENTS_WAR_AND_PEACE)
					GET_TEAM(GetID()).DoDeclareWar(eOriginatingPlayer, true, (TeamTypes)iI, /*bDefensivePact*/ true);
#else
					GET_TEAM(GetID()).DoDeclareWar((TeamTypes)iI, /*bDefensivePact*/ true);
#endif
				}
			}
		}
```

So the way it seems to be set up, the aggressor declares war on all DP and vassal civs when they declare war on another team; this makes sense to me.

What I'm curious about is that it passes on the variable bDefensivePact as true, which, if I'm understanding the function correctly, is passed on to DoPlayerDeclaredWarOnSomeone, and exempts the declaring player from diplomacy penalties.

In other words, you can evade the backstabbing penalty for declaring war on a friend by declaring war on someone they have a Defensive Pact with.

Why is this? When I originally read through DoPlayerDeclaredWarOnSomeone, I understood that (!bDefensivePact) was meant to make sure no penalties were given for declaring war as part of a Defensive Pact, because you weren't the aggressor, but it seems I was mistaken, since this seems to protect the aggressor against penalties for DPs that they were fully aware of before declaring.

Am I misunderstanding this, or was there a reason why this choice was made (e.g. the AI isn't factoring in DPs when declaring war)?

**Comment by HeathcliffWarriors** [comment]
@LoneGazebo Any thoughts on this?

**Comment by HeathcliffWarriors** [comment]
Hello again, I'm back from my vacation :)

Current status, these are the relevant code sections in **CvTeam.cpp** (look for DoDeclareWar and DoPlayerDeclaredWarOnSomeone).
```
		//are we an aggressor?
		//These only trigger on the 'first' major declaration - no chain declarations.
		if (bAggressor && !bDefensivePact)
		{
			//DPs and Vassal civs - we have to DOW them as well!
			for (iI = 0; iI < MAX_TEAMS; iI++)
			{
				if (!GET_TEAM((TeamTypes)iI).isAlive())
					continue;
				if (GET_TEAM((TeamTypes)iI).isMinorCiv())
					continue;

				//Defensive pacts and vassals trigger here.
				if (GET_TEAM((TeamTypes)iI).IsHasDefensivePact(eTeam) || GET_TEAM((TeamTypes)iI).IsVassal(eTeam))
				{
#if defined(MOD_EVENTS_WAR_AND_PEACE)
					GET_TEAM(GetID()).DoDeclareWar(eOriginatingPlayer, true, (TeamTypes)iI, /*bDefensivePact*/ true);
#else
					GET_TEAM(GetID()).DoDeclareWar((TeamTypes)iI, /*bDefensivePact*/ true);
#endif
				}
			}
		}

		//Secondary civs that will declare war along with us
		for (iI = 0; iI < MAX_TEAMS; iI++)
		{
			if (!GET_TEAM((TeamTypes)iI).isAlive())
				continue;
			if (GET_TEAM((TeamTypes)iI).isMinorCiv())
				continue;
				
			if (MOD_DIPLOMACY_CIV4_FEATURES && GET_TEAM((TeamTypes)iI).IsVassal(GetID()))
			{
#if defined(MOD_EVENTS_WAR_AND_PEACE)
				GET_TEAM((TeamTypes)iI).DoDeclareWar(eOriginatingPlayer, bAggressor, eTeam, /*bDefensivePact*/ true);
#else
				GET_TEAM((TeamTypes)iI).DoDeclareWar(eTeam, /*bDefensivePact*/ false);
#endif
			}
		}

		//Secondary check for vassals.
		for (iI = 0; iI < MAX_TEAMS; iI++)
		{
			if (!GET_TEAM((TeamTypes)iI).isAlive())
				continue;

			if (GET_TEAM((TeamTypes)iI).isMinorCiv())
				continue;

			if (MOD_DIPLOMACY_CIV4_FEATURES && GET_TEAM((TeamTypes)iI).IsVassal(eTeam))
			{
#if defined(MOD_EVENTS_WAR_AND_PEACE)
				GET_TEAM((TeamTypes)iI).DoDeclareWar(eOriginatingPlayer, bAggressor, GetID(), /*bDefensivePact*/ true);
#else
				GET_TEAM((TeamTypes)iI).DoDeclareWar(eTeam, /*bDefensivePact*/ false);
#endif
			}

			if (MOD_DIPLOMACY_CIV4_FEATURES && GET_TEAM((TeamTypes)iI).IsVassal(GetID()))
			{
#if defined(MOD_EVENTS_WAR_AND_PEACE)
				GET_TEAM((TeamTypes)iI).DoDeclareWar(eOriginatingPlayer, bAggressor, eTeam, /*bDefensivePact*/ true);
#else
				GET_TEAM((TeamTypes)iI).DoDeclareWar(eTeam, /*bDefensivePact*/ false);
#endif
			}
		}
	}

[...]

	int iMinorCivLoop;
	int iMajorCivLoop;

	// Minor Civs declaring war
	if (isMinorCiv())
	{

	}
	// Major is declaring War
	else
	{
		int iMajorCivLoop2;

		// Update what every Major Civ sees
		for (iMajorCivLoop = 0; iMajorCivLoop < MAX_MAJOR_CIVS; iMajorCivLoop++)
		{
			if (GET_PLAYER((PlayerTypes)iMajorCivLoop).getTeam() == GetID())
			{
				if (GET_PLAYER((PlayerTypes)iMajorCivLoop).isAlive())
				{
					// Alter Grand Strategy stats for all Majors in the game who've met this Major declaring war
					for (iMajorCivLoop2 = 0; iMajorCivLoop2 < MAX_MAJOR_CIVS; iMajorCivLoop2++)
					{
						// I don't care if it's me
						if (iMajorCivLoop != iMajorCivLoop2)
						{
							// Have I actually met this player declaring war?
							if (GET_TEAM(GET_PLAYER((PlayerTypes)iMajorCivLoop2).getTeam()).isHasMet(GET_PLAYER((PlayerTypes)iMajorCivLoop).getTeam()))
							{
								// Update Diplo stuff
#if defined(MOD_BALANCE_CORE)
								GET_PLAYER((PlayerTypes)iMajorCivLoop2).GetDiplomacyAI()->DoPlayerDeclaredWarOnSomeone((PlayerTypes)iMajorCivLoop, eTeam, bDefensivePact);
#else
								GET_PLAYER((PlayerTypes)iMajorCivLoop2).GetDiplomacyAI()->DoPlayerDeclaredWarOnSomeone((PlayerTypes)iMajorCivLoop, eTeam);
#endif
								if (!bDefensivePact)
								{
									// Major declaring war on Minor
									if (GET_TEAM(eTeam).isMinorCiv())
									{
										GET_PLAYER((PlayerTypes)iMajorCivLoop2).GetDiplomacyAI()->ChangeOtherPlayerNumMinorsAttacked((PlayerTypes)iMajorCivLoop, 1);
									}
									// Major declaring war on Major
									else
									{
										GET_PLAYER((PlayerTypes)iMajorCivLoop2).GetDiplomacyAI()->ChangeOtherPlayerNumMajorsAttacked((PlayerTypes)iMajorCivLoop, 1, eTeam);
									}
								}
							}
						}
					}
```

As far as I can tell, this is how things currently are:
- Player A, the originating player, decides to declare war on Player B.
- Player A also declares war on any of Player B's vassals and DPs; however, this sets the bDefensivePact flag for these DoWs, which eliminates many diplo penalties for declaring war.
- Player A's vassals declare war on Player B and their vassals/DPs; this also sets bDefensivePact

In DoPlayerDeclaredWarOnSomeone (**CvDiplomacyAI.cpp**), if bDefensivePact = true, it currently has the following effects if MOD_BALANCE_CORE_DIPLOMACY is defined:
- Ignores penalties for declaring war on a friend;
- Ignores penalties for breaking coop war agreements (added last time I made a pull request);
- Ignores penalties for breaking a military promise;
- Doesn't clear city liberation, civilians returned and other diplo bonuses;
- Ignores penalties for declaring war on a protected City-State.

It makes sense to me that vassals don't get penalized for DoW; what didn't make sense to me is that the aggressor apparently gets off scot-free for DoWing friends because it was done via a Defensive Pact - which the aggressor was fully aware of before making the declaration.

This is exploitable for the human player because rather than declare war on a friend, they can declare war on that friend's DP, and avoid the hefty diplo penalties.

I have, however, since given it some thought and have three hypotheses:
1) The intention of a Defensive Pact is that it's actually Player B declaring war on Player A, and therefore Player A shouldn't be penalized for it, even though the game code is set up in such a way to make Player A declare war on Player B;
2) It was too difficult for the AI to factor this in before a DoW, and they were getting hit with heavy diplo penalties;
3) Both.

I'm curious as to which of these three, if any, were the reason for this decision.

(NOTE: Originally the only penalties it didn't apply were those for DoWing a friend and protected CS; I added the other ones in previous pull requests because my original understanding was that bDefensivePact removed penalties from someone declaring war on the aggressor as part of a DP, and if they were to be granted immunity from the penalty for DoWing a friend, the others should reasonably apply as well)

(NOTE 2: DoPlayerDeclaredWarOnSomeone is a function I edited recently, as the function syntax was messy and unclear, but bDefensivePact has been set up this way for a long time, so it wasn't my doing)

**Comment by HeathcliffWarriors** [comment]
Anyway, while this is exploitable in that a player can avoid DoW penalties by declaring war on someone the player has a DP with, I think it's not *too* big an issue, since a DP is really the target AI agreeing to declare war on the aggressor.

**Comment by LoneGazebo** [comment]
Sorry this slipped through. You've thought about this quite a bit more than I, I defer to you on any solutions. BUT re: your hypotheses, 1 is correct.

**Comment by HeathcliffWarriors** [comment]
I figured that was the case. Mechanics-wise, I think it's okay after my changes.

However, while this doesn't change the actual mechanics and isn't high priority, it'd be nice if declaring war triggered a Defensive Pact in such a way that the other civ declares war on you, rather than you declaring war on them.

(i.e. You declare war on Player B; Player B has a Defensive Pact with Player C; "Player C has declared war on you!")

It's probably too much effort for something minor, however, so may as well leave it as-is.
